java-checks/src/main/java/org/sonar/java/checks/StandardFunctionalInterfaceCheck.java

```diff
@@ -0,0 +1,299 @@
+    registerInterface("java.util.function.Predicate<T>", "boolean", "T");
+    registerInterface("java.util.function.Supplier<T>", "T");
+    registerInterface("java.util.function.ToDoubleBiFunction<T,U>", "double", "T", "U");
+    registerInterface("java.util.function.ToDoubleFunction<T>", "double", "T");
+    registerInterface("java.util.function.ToIntBiFunction<T,U>", "int", "T", "U");
+    registerInterface("java.util.function.ToIntFunction<T>", "int", "T");
+    registerInterface("java.util.function.ToLongBiFunction<T,U>", "long", "T", "U");
+    registerInterface("java.util.function.ToLongFunction<T>", "long", "T");
+    registerInterface("java.util.function.UnaryOperator<T>", "T", "T");
+
+    // Each list of FunctionalInterface has to be sorted ascending by number of parametrized types so that smallest number
+    // of parametrized types take precedence. For example UnaryOperator<String> and Function<String,String> are equivalent,
+    // but UnaryOperator is preferred.
+    STD_INTERFACE_BY_PARAMETER_COUNT.values().forEach(list -> list.sort((a, b) -> Integer.compare(a.getGenericTypeCount(), b.getGenericTypeCount())));
+  }
+
+  private static void registerInterface(String name, String returnType, String... parameters) {
+    FunctionalInterface functionalInterface = new FunctionalInterface(name, returnType, parameters);
+    STD_INTERFACE_NAMES.add(functionalInterface.getName());
+    STD_INTERFACE_BY_PARAMETER_COUNT.computeIfAbsent(functionalInterface.getParameterCount(), key -> new ArrayList<>()).add(functionalInterface);
+  }
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return ImmutableList.of(Tree.Kind.INTERFACE);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    if (!hasSemantic()) {
+      return;
+    }
+    ClassTree classTree = (ClassTree) tree;
+    // classTree.simpleName() never null for Tree.Kind.INTERFACE
+    IdentifierTree issueLocation = classTree.simpleName();
+    Optional.of(classTree)
+      .filter(StandardFunctionalInterfaceCheck::isFunctionalInterface)
+      .filter(StandardFunctionalInterfaceCheck::isNonStandardFunctionalInterface)
+      .filter(StandardFunctionalInterfaceCheck::hasNoExtension)
+      .flatMap(StandardFunctionalInterfaceCheck::lookupFunctionalMethod)
+      .flatMap(StandardFunctionalInterfaceCheck::lookupMatchingStandardInterface)
+      .ifPresent(standardInterface -> reportIssue(issueLocation, buildIssueMessage(classTree, standardInterface)));
+  }
+
+  private static String buildIssueMessage(ClassTree interfaceTree, String standardInterface) {
+    if (interfaceTree.members().size() <= 1) {
+      return "Drop this interface in favor of \"" + standardInterface + "\".";
+    } else {
+      return "Make this interface extend \"" + standardInterface + "\" and remove the functional method declaration.";
+    }
+  }
+
+  private static Optional<MethodSymbol> lookupFunctionalMethod(ClassTree interfaceTree) {
```

Comment:
Could you please order the methods as they are used in the visitNode method. Will make the code much easier to read.